### PR TITLE
[DRAFT] Zero123 Improvements

### PR DESCRIPTION
**Pull request by voletiv**
Multiple improvements to Zero123 and other files:

- Sets good params for using Zero123! Doesn't use depth or normal images
- Changes background for Zero123 to non-trainable `solid-color-background` with white color
- Changes optimizer to Adam
- The normals are much smoother. Will affect details like thin fingers, etc. But works well overall, and is very conducive to Phase 2 using tetrahedra-sdf-grid
- Changes `zero123-prompt-processor` to `dummy-prompt-processor`
- Adds `resolution_milestones` to `data/image.py` : so the height and width can change with iters. For example in `zero123.yaml`:
```
data: # threestudio/data/image.py -> SingleImageDataModuleConfig
  image_path: ./load/images/hamburger_rgba.png
  height: [128, 256, 512]
  width: [128, 256, 512]
  resolution_milestones: [200, 300]
```
- Only loads depth and normal images if `lambda_depth` or `lambda_depth_rel` > 0, or `lambda_normal` > 0. Else doesn't load them, only loads the `_rgba` image. Also, resizes the images according to `resolution_milestones`
- Adds `random_aug` to `solid-colour-background` : uses a random background colour with probability `random_aug_prob`
- Sets bg_color as white, changed `ambient_ratio` from `0.1 + 0.9*rand` to `0.5 + 0.5*rand` : didn't notice any difference with this change
- TODO : fix max guidance_eval items to 4

GEOMETRY REFINEMENT using `tetrahedra-sdf-grid`
- Fixes to dataloaders to be compatible with nvdiffrast
- Adds scripts to refine Phase1 results using `tetrahedra-sdf-grid` : `configs/zero123-geometry.yaml` and `configs/experimental/imagecondition_refine.yaml`
- Adds `normal_consistency` loss during refinement
- Adds `laplacian_smoothness` loss during refinement : doesn't work yet, will fix later

https://github.com/threestudio-project/threestudio/assets/22424247/26ea37cb-36e4-4e07-a449-074472faa7b1

### CODE REVIEW

File: configs/experimental/imagecondition_zero123nerf.yaml [hunk 1]
```diff
@@ -95,16 +98,14 @@ system:
...
   guidance_type: "stable-diffusion-guidance"
   guidance:
     pretrained_model_name_or_path: "runwayml/stable-diffusion-v1-5"
-    guidance_scale: 7.5
-    min_step_percent: 0.2
-    # min_step_percent: [0, 0.66, 0.33, 2000]  # (start_iter, start_val, end_val, end_iter)
-    max_step_percent: 0.6
-    # max_step_percent: [0, 0.98, 0.66, 2000]
+    guidance_scale: 100
+    min_step_percent: [0, 0.4, 0.2, 200]  # (start_iter, start_val, end_val, end_iter)
```
Comment (claforte) [inline configs/experimental/imagecondition_zero123nerf.yaml:107]: Not for this PR, but for a later one... maybe try `min_step_percent: [0, 0.4, 0.02, 200]`... that might help with minor details

File: configs/experimental/imagecondition_zero123nerf.yaml [hunk 2]
```diff
@@ -131,33 +132,27 @@ system:
...
-    lambda_opaque: 0.05
+    lambda_normal_smooth: 10.0
+    lambda_3d_normal_smooth: 10.0
+    lambda_orient: 1.0
+    lambda_sparsity: 0.1
+    lambda_opaque: 0.1
 
   optimizer:
-    name: Adan
+    name: Adam
```
Comment (claforte) [inline configs/experimental/imagecondition_zero123nerf.yaml:148]: Did you notice any difference specifically tied to Adan vs Adam? Just curious.

File: threestudio/data/image.py [hunk 3]
```diff
@@ -29,8 +31,9 @@
 
 @dataclass
 class SingleImageDataModuleConfig:
-    height: int = 96
-    width: int = 96
+    height: Any = 96
+    width: Any = 96
```
Comment (claforte) [inline threestudio/data/image.py:35]: Instead of `height: Any`, it would be best to use `height:Union[int,List[int]]` or whatever types you expect. You might need to add a `from typing import Union` on top.

File: configs/experimental/imagecondition_zero123nerf.yaml (hunk 1 above)
Comment (voletiv) [reply configs/experimental/imagecondition_zero123nerf.yaml:107]: Sure!

File: configs/experimental/imagecondition_zero123nerf.yaml (hunk 2 above)
Comment (voletiv) [reply configs/experimental/imagecondition_zero123nerf.yaml:148]: Noticeably, no, but I didn't do a thorough ablation study as such. I wanted to stick to Adam with a simpler optimizer block, just because I'm more familiar with Adam than Adan.

File: threestudio/models/guidance/deep_floyd_guidance.py [hunk 4]
```diff
@@ -343,17 +343,17 @@ def guidance_eval(
...
+            :bs
+        ].unsqueeze(
             -1
         )  # sized [bs,50] > [bs,1]
         idxs = torch.min(large_enough_idxs, dim=1)[1]
         t = self.scheduler.timesteps_gpu[idxs]
 
         fracs = list((t / self.scheduler.config.num_train_timesteps).cpu().numpy())
-        imgs_noisy = latents_noisy.permute(0, 2, 3, 1)
+        imgs_noisy = (latents_noisy[:bs] / 2 + 0.5).permute(0, 2, 3, 1)
```
Comment (claforte) [inline threestudio/models/guidance/deep_floyd_guidance.py:363]: So the noisy_latents weren't scaled and biases properly for DeepFloyd, previously?

File: threestudio/models/guidance/zero123_guidance.py [hunk 5]
```diff
@@ -350,17 +350,17 @@ def guidance_eval(self, cond, t_orig, latents_noisy, noise_pred):
         # use only 50 timesteps, and find nearest of those to t
         self.scheduler.set_timesteps(50)
         self.scheduler.timesteps_gpu = self.scheduler.timesteps.to(self.device)
-        bs = latents_noisy.shape[0]  # batch size
-        large_enough_idxs = self.scheduler.timesteps_gpu.expand(
-            [bs, -1]
-        ) > t_orig.unsqueeze(
+        bs = min(4, latents_noisy.shape[0])  # batch size
```
Comment (claforte) [inline threestudio/models/guidance/zero123_guidance.py]: You're hardcoding the limit to 4, but in my PR merged yesterday, I made it configurable... please keep the configurability - I like to use 1 max_items, that makes it easier to compare vertically the results of different runs.

File: threestudio/systems/imagedreamfusion.py [hunk 6]
```diff
@@ -66,22 +67,22 @@ def training_substep(self, batch, batch_idx, guidance: str):
...
-            # is the batch item white? shaped [bs, 1]
-            is_white = (torch.rand(bs) > 0.5).to(self.device).float().unsqueeze(-1)
+            # # is the batch item white? shaped [bs, 1]
+            # is_white = (torch.rand(bs) > 0.5).to(self.device).float().unsqueeze(-1)
 
-            bg_color = bg_color * (1.0 - is_white) + white * is_white
+            # bg_color = bg_color * (1.0 - is_white) + white * is_white
 
-            ambient_ratio = 0.1 + 0.9 * random.random()
+            ambient_ratio = 0.5 + 0.5 * random.random()
```
Comment (claforte) [inline threestudio/systems/imagedreamfusion.py:85]: We should make this configurable and leave the default at `0.1 + 0.9 * random`... otherwise you might break some others' projects

File: threestudio/systems/imagedreamfusion.py [hunk 7]
```diff
@@ -363,3 +384,6 @@ def on_test_epoch_end(self):
             name="test",
             step=self.true_global_step,
         )
+        shutil.rmtree(
+            os.path.join(self.get_save_dir(), f"it{self.true_global_step}-test")
+        )
```
Comment (claforte) [inline threestudio/systems/imagedreamfusion.py:389]: Are you sure this doesn't mess up the wandb upload? It happens asynchronously IIRC.
IMHO it would be best to add an option to delete most of the output folder for the run, after Trainer() is completely done training.

File: threestudio/systems/zero123.py [hunk 8]
```diff
@@ -67,20 +68,21 @@ def training_substep(self, batch, batch_idx, guidance: str):
...
+            # bg_color = torch.rand(bs, 3).to(self.device)  # claforte: use dtype
 
-            # Override 50% of the bgcolors with white.
-            # This results in better predictions with Zero123,
-            # since it was trained with a constant white background.
-            white = torch.ones(bs, 3).to(self.device)
+            # # Override 50% of the bgcolors with white.
+            # # This results in better predictions with Zero123,
+            # # since it was trained with a constant white background.
+            # white = torch.ones(bs, 3).to(self.device)
```
Comment (claforte) [inline threestudio/systems/zero123.py:77]: You disabled the "replace bg color by white 50% of the time"... but are you sure that this improves convergence?

File: threestudio/systems/zero123.py [hunk 9]
```diff
@@ -367,3 +399,6 @@ def on_test_epoch_end(self):
             name="test",
             step=self.true_global_step,
         )
+        shutil.rmtree(
+            os.path.join(self.get_save_dir(), f"it{self.true_global_step}-test")
+        )
```
Comment (claforte) [inline threestudio/systems/zero123.py:404]: same as previous comment

**Review by claforte — CHANGES_REQUESTED** [review]
Good work Vikram... please address these comments...

File: threestudio/models/guidance/deep_floyd_guidance.py (hunk 4 above)
Comment (voletiv) [reply threestudio/models/guidance/deep_floyd_guidance.py:363]: Yes, none of the latents were. This is only during guidance_eval

File: threestudio/models/guidance/zero123_guidance.py (hunk 5 above)
Comment (voletiv) [reply threestudio/models/guidance/zero123_guidance.py]: I've made it configurable now, please check!

File: threestudio/systems/imagedreamfusion.py (hunk 7 above)
Comment (voletiv) [reply threestudio/systems/imagedreamfusion.py:389]: it doesn't affect it. We do this for `on_validation_epoch_end` as well. Without it, in addition to test.mp4, there also remains an extra directory with all the images used to make test.mp4

File: threestudio/systems/zero123.py (hunk 8 above)
Comment (voletiv) [reply threestudio/systems/zero123.py:77]: In our case, it doesn't affect the background at all because our bg is always white.
I can re-enable it if you think that's better?

File: threestudio/systems/zero123.py (hunk 9 above)
Comment (voletiv) [reply threestudio/systems/zero123.py:404]: Same as previous reply ;)

File: threestudio/systems/imagedreamfusion.py (hunk 6 above)
Comment (voletiv) [reply threestudio/systems/imagedreamfusion.py:85]: Sure, I'll revert it back to 0.1 + 0.9*rand for now.
I didn't see any noticeable difference between the two cases though.

File: threestudio/data/image.py (hunk 3 above)
Comment (voletiv) [reply threestudio/data/image.py:35]: Ideally yes, but OmegaConf does not support Union of containers. Will mention this in a comment

**Review by claforte — DISMISSED** [review]

**Review by claforte — APPROVED** [review]
